From e17477bb4cf89d6551975556c0823695a199388d Mon Sep 17 00:00:00 2001
From: Nathaniel Clark <nathaniel.l.clark@intel.com>
Date: Thu, 31 Mar 2016 11:12:38 -0400
Subject: [PATCH] LU-7863 osd-zfs: dmu_prefetch change in ZFS master

In ZFS master (post zfs-0.6.5-release) dmu_prefetch now has 6
arguments instead of 4.  This accounts for the new style dmu_prefetch,
while keeping backward compatibility (back to 0.6.2 which initially
exported dmu_prefetch).

Signed-off-by: Nathaniel Clark <nathaniel.l.clark@intel.com>
Change-Id: Ia06e5605b06ef1f51817418183e479d591231e5a
---

diff --git a/config/lustre-build-zfs.m4 b/config/lustre-build-zfs.m4
index 0d45ce4..07856e9 100644
--- a/config/lustre-build-zfs.m4
+++ b/config/lustre-build-zfs.m4
@@ -440,6 +440,17 @@
 			AC_DEFINE(HAVE_SPA_MAXBLOCKSIZE, 1,
 				[Have spa_maxblocksize in ZFS])
 		])
+		LB_CHECK_COMPILE([if ZFS has 'dmu_prefetch' with 6 args],
+		dmu_prefetch, [
+			#include <sys/dmu.h>
+		],[
+			objset_t *os = NULL;
+
+			dmu_prefetch(os, 0, 0, 0, 0, 0);
+		],[
+			AC_DEFINE(HAVE_DMU_PREFETCH_6ARG, 1,
+				[Have 6 Argument dmu_pretch in ZFS])
+		])
 	])
 
 	AM_CONDITIONAL(ZFS_ENABLED, [test "x$enable_zfs" = xyes])
diff --git a/lustre/osd-zfs/osd_index.c b/lustre/osd-zfs/osd_index.c
index 705c86d..d476435 100644
--- a/lustre/osd-zfs/osd_index.c
+++ b/lustre/osd-zfs/osd_index.c
@@ -1477,7 +1477,7 @@
 		/* dmu_prefetch() was exported in 0.6.2, if you use with
 		 * an older release, just comment it out - this is an
 		 * optimization */
-		dmu_prefetch(dev->od_os, it->mit_prefetched_dnode, 0, 0);
+		osd_dmu_prefetch(dev->od_os, it->mit_prefetched_dnode);
 
 		it->mit_prefetched++;
 	}
diff --git a/lustre/osd-zfs/osd_internal.h b/lustre/osd-zfs/osd_internal.h
index fa0c3f2..b3cb2e7 100644
--- a/lustre/osd-zfs/osd_internal.h
+++ b/lustre/osd-zfs/osd_internal.h
@@ -593,4 +593,16 @@
 #define	osd_zio_buf_free(buf, size)	zio_buf_free(buf, size)
 #endif
 
+#ifdef HAVE_DMU_PREFETCH_6ARG
+static inline void osd_dmu_prefetch(objset_t *os, uint64_t object)
+{
+	dmu_prefetch(os, object, 0, 0, 0, ZIO_PRIORITY_ASYNC_READ);
+}
+#else
+static inline void osd_dmu_prefetch(objset_t *os, uint64_t object)
+{
+	dmu_prefetch(os, object, 0, 0);
+}
+#endif
+
 #endif /* _OSD_INTERNAL_H */
diff --git a/lustre/osd-zfs/osd_oi.c b/lustre/osd-zfs/osd_oi.c
index ef9aefc..7752588 100644
--- a/lustre/osd-zfs/osd_oi.c
+++ b/lustre/osd-zfs/osd_oi.c
@@ -488,7 +488,7 @@
 	}
 
 	if (rc == 0)
-		dmu_prefetch(dev->od_os, *oid, 0, 0);
+		osd_dmu_prefetch(dev->od_os, *oid);
 
 	RETURN(rc);
 }
